src/driver-sinowealth.c

```diff
@@ -1373,140 +1370,37 @@ sinowealth_update_macro_from_action(struct ratbag_profile *profile, struct ratba
-		 */
-		.button_count = 5 + 9,
-	},
-
-	/* 258a:0033 devices. */
-	{
-		.fw_version = "V102",
-		.device_name = "Glorious Model D",
-		.led_type = LED_RBG,
-		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
-		.vid = 0x258a,
-		.pid = 0x33,
-		.button_count = 6,
-	},
-
-	/* 258a:0036 devices. */
-	{
-		.fw_version = "V103",
-		.device_name = "Glorious Model O/O- (updated firmware)",
-		.led_type = LED_RBG,
-		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
-		.vid = 0x258a,
-		.pid = 0x36,
-		.button_count = 6,
-	},
-
-	/* 258a:003E devices. */
-	{
-		/* Untested. */
-		/* This matches both Classic and Ace versions. */
-		.fw_version = "3105",
-		.device_name = "G-Wolves Hati HT-S Wired",
-		.led_type = LED_NONE,
-		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
-		.vid = 0x258a,
-		.pid = 0x3e,
-		.button_count = 6,
-	},
-
-	/* 2581:1007 devices. */
-	{
-		.fw_version = "9677",
-		.device_name = "Marvo Scorpion G961",
-		.led_type = LED_RGB,
-		/* Official software allows up to 12000 DPI instead of 10200. */
-		.sensor_type = SINOWEALTH_SENSOR_PMW3327,
-		.vid = 0x258a,
-		.pid = 0x1007,
-		.button_count = 6,
-	},
-
-	/* Dummy device. Make sure it's the last entry. */
-	{
-		.fw_version = "0000",
-		.device_name = "Generic SinoWealth Mouse",
-		.led_type = LED_RBG,
-		.sensor_type = SINOWEALTH_SENSOR_PMW3360,
-		.vid = 0xffff,
-		.pid = 0xffff,
-		.button_count = 6,
-	},
-};
 
+// @return Supported device data for the device or `NULL`.
 static const struct sinowealth_device_data *
-sinowealth_find_device_data(struct input_id ids, const char *fw_version)
+sinowealth_find_device_data(struct ratbag_device *device, const char *fw_version)
 {
-	const struct sinowealth_device_data *device_data = NULL;
+	const struct ratbag_device_data *data = device->data;
+
+	const struct list *supported_devices = ratbag_device_data_sinowealth_get_supported_devices(data);
 
-	ARRAY_FOR_EACH(sinowealth_supported_devices, device_data) {
-		if (device_data->vid != ids.vendor || device_data->pid != ids.product)
+	struct sinowealth_device_data *device_data = NULL;
+	list_for_each(device_data, supported_devices, link) {
+		if (device_data->fw_version == NULL || device_data->device_name == NULL) {
+			log_error(device->ratbag, "Skipping invalid device data\n");
 			continue;
+		}
 
-		if (!strneq(fw_version, device_data->fw_version, sizeof(device_data->fw_version)))
+		if (!strneq(fw_version, device_data->fw_version, SINOWEALTH_FW_VERSION_LEN))
 			continue;
 
 		return device_data;
 	}
 
-	/* Return the last entry in `sinowealth_device_data` - a dummy. */
-	return device_data;
+	const struct sinowealth_device_data *fallback_device_data = ratbag_device_data_sinowealth_get_fallback_device(data);
+	if (fallback_device_data != NULL) {
+		log_info(device->ratbag, "Your mouse is not supported officially, but still seems to work fine.\n");
+		log_info(device->ratbag, "Consider reporting model name and above firmware version to make it official.\n");
```

Comment:
I wouldn't use the term `officially` here, since we're not sinowealth. How about something like:
```
This device is not known to libratbag, using fallback behavior. Please report model name and above firmware version to the libratbag project.
```
Or something along these lines anyway.

I'd also be careful with the term `supported` since we don't really support anything. We know of it and maybe find time to help fix it, but that's about it :)